# <span style="color:blue"> 이토록 쉬운 통계 & R </span>

<img src="../image/cover.png" width="200" height="400" align="left">  임경덕 지음

## PART 4 데이터 분석 도구, R

## 90% 순서대로 살펴보는 BR31-6 (36일차)

   * 데이타프로젝트 (http://dataproject.xyz) 또는 (http://cafe.naver.com/dataartproject) 에 미리 저장된 스크립트를 다운 받을 수 있습니다.
   * 스크립트를 활용하면 직접 명령어를 입력하지 않고 저장된 명령어들을 실행할 수 있습니다.
   * 지금 설치된 R 버젼은 3.4.3 입니다. 제목 BR31 은 Basic R 31 로 3.1 버젼입니다. 그러나 큰 차이는 없을 것 같습니다.
   * page 282 subset() 함수부터 계속 진행

### 데이터 파일은 dataartproject.xyz 에 있습니다.

In [8]:
# subway 데이터 로드
subway = read.csv("Practice/data/subway.csv")

In [9]:
# 변수명 조정
names(subway) = c("Station", "Date", "InOut", paste0("H", 5:25))

### subset() 함수로 부분 데이터 선택하기

In [10]:
subdata1 = subset(subway, Station %in% c("강남(222)", "명동(424)"))

In [11]:
# 데이터 확인하기
str(subdata1)

'data.frame':	124 obs. of  24 variables:
 $ Station: Factor w/ 119 levels "가락시장(340)",..: 2 2 2 2 2 2 2 2 2 2 ...
 $ Date   : Factor w/ 31 levels "2016-12-01","2016-12-02",..: 1 1 2 2 3 3 4 4 5 5 ...
 $ InOut  : Factor w/ 2 levels "승차","하차": 1 2 1 2 1 2 1 2 1 2 ...
 $ H5     : num  421 520 420 509 903 320 767 146 373 509 ...
 $ H6     : num  807 2442 806 2377 907 ...
 $ H7     : num  2087 6944 1952 6687 809 ...
 $ H8     : num  3435 15612 3357 15203 1228 ...
 $ H9     : num  2619 13090 2511 12988 1687 ...
 $ H10    : num  2343 6066 2298 6256 2248 ...
 $ H11    : num  2952 4568 3035 5096 3218 ...
 $ H12    : num  3356 4340 3931 4478 4472 ...
 $ H13    : num  4835 5396 4957 5805 5693 ...
 $ H14    : num  4740 4784 5229 5293 6226 ...
 $ H15    : num  5876 4559 6116 5061 7071 ...
 $ H16    : num  6389 4619 6723 5432 8459 ...
 $ H17    : num  8641 6582 9256 7398 9650 ...
 $ H18    : num  14714 9701 15777 11822 7618 ...
 $ H19    : num  10351 8178 11210 13295 5721 ...
 $ H20    : num  7976 3

In [13]:
subdata2 = subset(subway, Station %in% c("강남(222)", "명동(424)") & Date=="2016-12-24")

In [14]:
str(subdata2)

'data.frame':	4 obs. of  24 variables:
 $ Station: Factor w/ 119 levels "가락시장(340)",..: 2 2 37 37
 $ Date   : Factor w/ 31 levels "2016-12-01","2016-12-02",..: 24 24 24 24
 $ InOut  : Factor w/ 2 levels "승차","하차": 1 2 1 2
 $ H5     : num  1133 356 58 33
 $ H6     : num  1093 1254 112 529
 $ H7     : num  760 1264 183 831
 $ H8     : num  1056 2993 386 1281
 $ H9     : num  1471 6211 765 2472
 $ H10    : num  2113 5394 1527 3306
 $ H11    : num  3176 5733 1709 4276
 $ H12    : num  4460 6098 2074 5690
 $ H13    : num  5509 7159 2483 6636
 $ H14    : num  5883 6537 3025 7463
 $ H15    : num  6573 6079 3806 7285
 $ H16    : num  7672 5995 4742 8065
 $ H17    : num  8170 6118 5802 8345
 $ H18    : num  6446 6693 6164 7868
 $ H19    : num  4844 5103 6268 6047
 $ H20    : num  4667 4303 6423 4118
 $ H21    : num  5829 4115 7093 3013
 $ H22    : num  6575 3179 6998 1748
 $ H23    : num  3848 2204 3818 796
 $ H24    : num  100 254 3 79
 $ H25    : num  0 0 0 0


   * subdata1 은 강남, 명동 두 역의 데이터 124건의 관측치에 대한 데이타입니다.
   * sundata2 는 강남, 명동 두역의 2016년 12월 24일 데이터 4건에 대한 데이터입니다.
   
   
   
   * 만일 밤 11 시 이후의 데이터를 구하기 위해선 select 옵션을 사용해서 다음과 같이 선택합니다.

In [15]:
subdata3 = subset(subway, Station %in% c("강남(222)", "명동(424)") & Date=="2016-12-24", select=c(Station, InOut, H23, H24, H25))

In [16]:
subdata3

Station   InOut H23  H24 H25
1969 강남(222) 승차  3848 100 0  
1970 강남(222) 하차  2204 254 0  
6743 명동(424) 승차  3818   3 0  
6744 명동(424) 하차   796  79 0

### dim()/nrow()/length() 로 변수/데이터 크기 확인 하기

In [18]:
dim(subdata1)

[1] 124  24

In [19]:
nrow(subdata1)

[1] 124

In [20]:
length(subway$Station)

[1] 7378

In [21]:
length(levels(subway$Station))

[1] 119

   * subdata1 에는 124 개의 관측치 와 24개의 변수가 있습니다.
   * Subway 의  Station 관측치를 세면 총 7378 개의 관측치가 있습니다.
   * 수준을 count 하려면 levels 함수를 이용하여 다음 length 를 구하면 119 개의 Station 이 있습니다.

### write.csv() 함수를 이용하여 csv 파일로 저장하기

In [22]:
# 컬럼 header 를 포함해서
write.csv(subdata3, "Practice/data/result.csv")

In [23]:
# 컬럼 header 를 뺴고
write.csv(subdata3, "Practice/data/result.csv", row.names=FALSE)

### install.packages() 로 패키지 설치 하기

In [24]:
install.packages("ca")


The downloaded binary packages are in
	/var/folders/3b/1f9v1lwn73q8hwwj6344j9pw0000gn/T//RtmpISUDLk/downloaded_packages


### library() 함수로 패키지 불러오기

In [26]:
# 설치한 패키지를 사용하려면 library 함수를 이용하여 패키지를 불러야 합니다.
library(ca)

In [27]:
# 설치하지 않은 패키지를 불룰경우에는 에러 메세지가 나옵니다.
library(cza)

ERROR: Error in library(cza): there is no package called ‘cza’


In [28]:
# ca 패키지를 불렀으므로 데이터를 확인 할 수 있습니다.
smoke

none light medium heavy
SM  4    2     3      2   
JM  4    3     7      4   
SE 25   10    12      4   
JE 18   24    33     13   
SC 10    6     7      2

지금 까지 31 개의 기본적인 R 연산자, 함수 기능들을 살펴 보았습니다. 다음 부터는 실제 데이터를 이용하여 다양한 분석 예제들을 직접 실행해 보면서 살펴 보겠습니다.